[SPARK-17018][SQL] literals.sql for testing literal parsing

petermaxlee opened the pull request:
## What changes were proposed in this pull request?

This patch adds literals.sql for testing literal parsing end-to-end in SQL.
## How was this patch tested?

The patch itself is only about adding test cases.

petermaxlee committed:
[SPARK-17018][SQL] literals.sql for testing literal parsing

petermaxlee commented on sql/core/src/test/resources/sql-tests/results/literals.sql.out:
```diff
@@ -0,0 +1,136 @@
+
+
+-- !query 5
+select 32768 S
```
Shouldn't this throw an exception?

cc @cloud-fan / @rxin / @hvanhovell

hvanhovell replied:
Why? This will create Integer literal `32768` aliased as `S`. `select 32768S` does throw an exception.

petermaxlee commented on sql/core/src/test/resources/sql-tests/results/literals.sql.out:
```diff
@@ -0,0 +1,136 @@
+
+
+-- !query 7
+select 9223372036854775808 L
```
Shouldn't this throw an exception since 9223372036854775808 is larger than the valid range?

cc @cloud-fan / @rxin / @hvanhovell

petermaxlee replied:
Actually my fault - let me update it since no space is allowed.

hvanhovell replied:
See my previous comment on shorts. The same thing. This will produce a BigDecimal with an Alias.

petermaxlee commented on sql/core/src/test/resources/sql-tests/results/literals.sql.out:
```diff
@@ -0,0 +1,136 @@
+
+
+-- !query 11
+select 9223372036854775807, -9223372036854775808
```
this is a bug reported in SPARK-17013

SparkQA commented:
**[Test build #63593 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63593/consoleFull)** for PR 14598 at commit [`15d2eec`](https://github.com/apache/spark/commit/15d2eecc1702b1556eadbabcf530c207fa64d8ee).

petermaxlee committed:
Add more floating point literals

petermaxlee commented on sql/core/src/test/resources/sql-tests/results/literals.sql.out:
```diff
@@ -0,0 +1,136 @@
+
+
+-- !query 13
+select 1D, 1 D, 1.2D, 1e10, 1.5e5, .10 D, 0.10 D, .1e5
```
there is a bug here too.

I would expect .10 D to be parsed as double, not decimal.

hvanhovell replied:
`BD` is a bigdecimal and `D` is a double.

SparkQA commented:
**[Test build #63596 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63596/consoleFull)** for PR 14598 at commit [`c40c957`](https://github.com/apache/spark/commit/c40c9576683b1da01814ccba6a2347ea22011f50).

petermaxlee committed:
remove spaces

petermaxlee commented on sql/core/src/test/resources/sql-tests/results/literals.sql.out:
```diff
@@ -0,0 +1,148 @@
+-- !query 7 schema
+struct<>
+-- !query 7 output
+org.apache.spark.sql.catalyst.parser.ParseException
```
This exception message can be better. It doesn't actually say out of range.

hvanhovell replied:
Hmmmm - this is funny. The exception/message is produced by `java.lang.Long.parseLong(...)`, but that doesn't seem to produce something sensible. I was expecting the something similar to the error `java.lang.Short.parseShort(...)` produces.

petermaxlee replied:
Should we parse integral literals as BigInteger, and then turn them into appropriate types? That way we have more control.

hvanhovell replied:
We do that already for 'untyped' (without a suffix) integral literals. I like your suggestion (this means we can also control the exception for Short better), could you open an PR for this?

petermaxlee replied:
Sure. Will do.

hvanhovell replied:
Nevermind this one. We have someone working on it.

petermaxlee commented on sql/core/src/test/resources/sql-tests/results/literals.sql.out:
```diff
@@ -0,0 +1,148 @@
+-- !query 5 output
+org.apache.spark.sql.catalyst.parser.ParseException
+
+Value out of range. Value:"32768" Radix:10(line 1, pos 7)
```
This exception message is a little bit weird with "radix".

hvanhovell replied:
It is the error you get when calling `java.lang.Short.valueOf`. I do agree that the radix info is not that useful when you already assume this `10`.

SparkQA commented:
**[Test build #63597 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63597/consoleFull)** for PR 14598 at commit [`243cd39`](https://github.com/apache/spark/commit/243cd390259e07c06fa58a6f0a6c7994217e64da).

petermaxlee committed:
Large decimals

SparkQA commented:
**[Test build #63598 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63598/consoleFull)** for PR 14598 at commit [`d60b2bb`](https://github.com/apache/spark/commit/d60b2bbef1fc00c8dfe0872a6ba22799b0067b06).

SparkQA commented:
**[Test build #63593 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63593/consoleFull)** for PR 14598 at commit [`15d2eec`](https://github.com/apache/spark/commit/15d2eecc1702b1556eadbabcf530c207fa64d8ee).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63593/
Test PASSed.

SparkQA commented:
**[Test build #63596 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63596/consoleFull)** for PR 14598 at commit [`c40c957`](https://github.com/apache/spark/commit/c40c9576683b1da01814ccba6a2347ea22011f50).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63596/
Test PASSed.

SparkQA commented:
**[Test build #63598 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63598/consoleFull)** for PR 14598 at commit [`d60b2bb`](https://github.com/apache/spark/commit/d60b2bbef1fc00c8dfe0872a6ba22799b0067b06).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63598/
Test FAILed.

SparkQA commented:
**[Test build #63597 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63597/consoleFull)** for PR 14598 at commit [`243cd39`](https://github.com/apache/spark/commit/243cd390259e07c06fa58a6f0a6c7994217e64da).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63597/
Test PASSed.

hvanhovell commented on sql/core/src/test/resources/sql-tests/inputs/literals.sql:
```diff
@@ -0,0 +1,52 @@
+select 1234567890123456789012345678901234567890.0;
+
+-- super large scientific notation numbers should still be valid doubles
+select 123456789012345678901234567890123456789e10, 123456789012345678901234567890123456789.1e10;
```
Shouldn't we also add really large double, `1E309` for instance (that will actually evaluate to positive infinity).

petermaxlee replied:
Let me add that.

hvanhovell commented:
This is pretty cool :)

I am comparing this to the `ExpressionParserSuite`. Shouldn't we add support for more complex string cases, intervals and a type constructors, see: https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala#L331-L498. If we do we can move these tests out of the `ExpressionParserSuite`.

petermaxlee committed:
Merge branch 'master' into SPARK-17018-2

petermaxlee committed:
Add more literals

SparkQA commented:
**[Test build #63625 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63625/consoleFull)** for PR 14598 at commit [`00bc63a`](https://github.com/apache/spark/commit/00bc63aa0686ecd2e3bb4df63c9bf7c1c0124efd).

petermaxlee commented:
I have updated this to include more string literals and added timestmap/date/interval parsing. That said, I didn't add all the test cases for interval because there were a large number, and I felt those are best left for parser unit tests.

petermaxlee committed:
remove \b

petermaxlee committed:
Remove \0

petermaxlee commented:
I also didn't include \b and \0 parsing. Otherwise github shows the result file as binary and refuse to display the diff, which makes it more difficult to review.

SparkQA commented:
**[Test build #63626 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63626/consoleFull)** for PR 14598 at commit [`5565144`](https://github.com/apache/spark/commit/55651444f799d3418bcb5c4fc93ada4af7d3b9f7).

hvanhovell commented on sql/core/src/test/resources/sql-tests/inputs/literals.sql:
```diff
@@ -0,0 +1,92 @@
+-- invalid timestamp
+select timestamp '2016-33-11 20:54:00.000';
+
+-- internal
```
NIT: interval? :)

SparkQA commented:
**[Test build #63625 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63625/consoleFull)** for PR 14598 at commit [`00bc63a`](https://github.com/apache/spark/commit/00bc63aa0686ecd2e3bb4df63c9bf7c1c0124efd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63625/
Test PASSed.

SparkQA commented:
**[Test build #63626 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63626/consoleFull)** for PR 14598 at commit [`5565144`](https://github.com/apache/spark/commit/55651444f799d3418bcb5c4fc93ada4af7d3b9f7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63626/
Test PASSed.

petermaxlee committed:
Fix typo

rxin commented:
I'm going to merge this in master/2.0.
